Fix: Fail early when database cluster does not respond

amotl opened the pull request:
## Problem

@shraik started using sqlalchemy-cratedb and reported that its behaviour deviates from other vendors by not failing on `engine.connect()` when the database server is not available.

- https://github.com/crate/sqlalchemy-cratedb/issues/218

We found this is not actually on the SQLAlchemy dialect, but on the DBAPI driver already, which exhibits the same behaviour.

## Solution

The patch extends the `_lowest_server_version` method to re-raise ~the last `ConnectionError` when no connection can be made to any configured server node~ a `ConnectionError` when connecting to all server nodes fails, including all error messages.

By doing it this way, we didn't need to submit a dummy SQL command like originally planned. We think it is much better this way because it does not pollute the server-side statement log.

## Details
As an additional benefit, the software tests in `test_connection.py` are now actual integration tests.

/cc @mfussenegger, @seut, @surister

amotl committed:
Error handling: Fail early when database cluster does not respond

kneth commented on src/crate/client/connection.py:
```diff
@@ -197,14 +198,20 @@ def get_blob_container(self, container_name):
-            except (ValueError, ConnectionError):
+            except ConnectionError as ex:
+                last_connection_error = ex
+                continue
```
Won't this `continue` let you try all active servers? Do you want to stop at the first exception?

amotl replied:
Yes and no.

kneth replied:
I probably take the change log entry too literally.

matriv replied:
Sorry, so what's happening? does it retry all servers provided and then throw the last connection error? (shouldn't it throw the first connection error maybe?)

seut replied:
If it would throw the first error, the whole logic of supporting multiple servers would be invalidated I think.

seut replied:
ah sry got this wrong, not sure if throwing the first error would be better. both could make sense imho.

seut replied:
another option could be to put all connection errors into an exception if all servers raise connection errors.

amotl replied:
> If it would throw the first error, the whole logic of supporting multiple servers would be invalidated I think.

I share the same opinion that it would defeat the whole purpose, right.

amotl replied:
@matriv: Do you like it better this way, now with 306feb5869791e?

amotl commented on docs/by-example/client.rst:
```diff
@@ -35,12 +35,8 @@ respond, the request is automatically routed to the next server:
-    >>> connection.client._active_servers
-    ['http://127.0.0.1:4200']
-    >>> connection.close()
+If no ``servers`` are supplied to the ``connect`` method, the default address
+``http://127.0.0.1:4200`` is used.
```
There is no server at `http://127.0.0.1:4200`. This test case just didn't fail because `connect()` did not raise an exception up until now.

Now, there is no longer a way to validate connecting to the default address per doctest file, because there is no server listening at `http://127.0.0.1:4200`. Because the core information is still viable, all what's left is pure prose, rephrased a bit.

amotl commented on CHANGES.rst:
```diff
@@ -5,6 +5,9 @@ Changes for crate
 Unreleased
 ==========
 
+- Changed connection behaviour to fail early when database server
+  does not respond
```
@kneth: Maybe rephrase this statement like this?

```suggestion
- Changed connection behaviour to fail early when database cluster
  does not respond
```

surister reviewed (approved):
LGTM

seut reviewed (approved):
Lgtm

amotl committed:
Error handling: Only re-raise ConnectionError when all servers fail

The procedure will collect all `ConnectionError` instances and include
them into the exception message of the final `ConnectionError`.

amotl commented on src/crate/client/connection.py:
```diff
@@ -197,14 +198,21 @@ def get_blob_container(self, container_name):
                 continue
             if not lowest or version < lowest:
                 lowest = version
+        if lowest is None and len(connection_errors) == server_count:
+            raise ConnectionError(str(connection_errors))
```
`str(connection_errors)` uses an opaque way, just serializing the list of exception instances into a string, the Python way. Do you have any objections or suggestions to do it differently?

amotl replied:
> [...] suggestions to do it differently, like using JSON instead?

The procedure now uses JSON per e818221ac.

amotl committed:
Error handling: Use JSON for bundling multiple ConnectionError instances

All `ConnectionError` instances that have been collected will be
serialized into JSON now.

seut reviewed (approved):
👍

amotl commented:
We just published a pre-release package including those updates per [crate-2.1.0.dev0](https://pypi.org/project/crate/2.1.0.dev0/).

matriv reviewed (commented):
@amotl Thank you! Is there a way to test that we don't fail the connection attempt, but only if only all servers provided fail?
Can we also test the exception when multiple nodes fail?

surister commented on src/crate/client/connection.py:
```diff
@@ -197,14 +199,21 @@ def get_blob_container(self, container_name):
             if not lowest or version < lowest:
                 lowest = version
+        if connection_errors and len(connection_errors) == server_count:
+            raise ConnectionError(json.dumps(list(map(str, connection_errors))))
```
why not just `raise ConnectionError(error_list)`?

amotl replied:
I've elaborated about it [here](https://github.com/crate/crate-python/pull/711#discussion_r2081480875), but I am also not sure, that's why I am also asking about your opinion.

amotl commented:
Hi @surister. The tests are failing now after rebasing on top of the recent test suite refactorings. Can I humbly ask you to look into it? Maybe you can spot the flaw quickly on the interface of both?

surister commented:
> Hi @surister. The tests are failing now after rebasing on top of the recent test suite refactorings. Can I humbly ask you to look into it? Maybe you can spot the flaw quickly on the interface of both?

looking

amotl commented:
Hi. I lost a bit track of this, maybe @surister or others can elaborate better. Could you also add tackling this stale patch to your backlog, @bgunebakan, when applicable?

bgunebakan committed:
Merge branch 'main' into fail-on-connect

bgunebakan committed:
fix linter errors

bgunebakan committed:
Fix test suite compatibility with fail-on-connect behavior change

bgunebakan committed:
Improve error handling in _lowest_server_version and add tests for connection errors

bgunebakan committed:
Update docs with new changes

bgunebakan commented:
@kneth @matriv @seut The PR was stale for a while, I updated the branch with main and fix did improvements on tests, log handling and docs. I kindly ask you to review again. Thanks in advance.

seut commented on tests/client/test_connection.py:
```diff
@@ -12,6 +13,40 @@
+
+
+def test_invalid_server_address():
+    client = Client(servers="localhost:4202")
```
Some other (manually started) node may listen on`4202`.
Let's use a complete different port to be safe ;) 

```suggestion
    client = Client(servers="localhost:1234")
```

seut commented on CHANGES.rst:
```diff
@@ -2,6 +2,13 @@
+
+- Breaking change: ``connect()`` now raises ``ConnectionError`` immediately if
+no configured server node responds. Changed connection behaviour to fail early 
+if the database cluster does not respond.
```
Isn't the 2nd sentence repeating the first in some kind?

```suggestion
- Breaking change: ``connect()`` now raises ``ConnectionError`` immediately if
no configured server node responds.
```

seut reviewed (approved):
just some minor suggestions.

bgunebakan committed:
Apply suggestions from code review

Co-authored-by: Sebastian Utz <su@rtme.net>
